### DISCUSSION TITLE
how to auto detect device name while settingup 2fa for user

### DISCUSSION THREAD

**Discussion by sangameshwar550** [keycloak/keycloak#19836] [Q&A]
Hi All,
I am looking to Auto detect the scanned device name programmatically while doing 2fa setup ex: Phone 11 or Oneplus nord 2, etc..., 
is there any way provided by keycloak to do so
Thanks in advance

**Comment by jbman** [comment]
If you mean TOTP authenticator app based 2FA: The QR coode only provides data from Keycloak to the authenticator app. Nothing is sent back to Keycloak.

**Reply by phenomuser** [reply]
Hi Johannes Bergmann,

Thanks for the replay, Yeah  I was asking about TOTP based 2FA, 
Is there any way or any 3rd party Keycloak plugin to auto detect the TOTP authenticator device name programetically.

Regards,
Sanga

**Reply by darius-m** [reply]
No, I doubt it's possible. As @jbman mentioned, the OTP applications do not send anything back to the Keycloak server, as it is not required, so Keycloak only ever gets the user's input.

TOTP is an extension of the HOTP specification that uses the current time instead of a number of clicks - it basically runs `hash(str(int(current_time / 30s)) + secret_salt)` to generate the OTP. Both the server and the application know the secret salt, and the current time is readily available. This is why it's necessary to use NTP (or other time sync protocols) on both the server and the device that generates the codes; otherwise, the codes that are generated by the server and the device will differ even if the same salt is used. To account for time skews, a look-around window is used (usually one or two intervals), so more than one code is accepted. However, since there is not communication between the application and Keycloak, large time skews will result in invalid codes, without the option to automatically adjust in case of an error.

**Reply by sangameshwar550** [reply]
HI @darius-m , thanks for the replay, 
Since its not possible, we are going with manually asking user to provide the device name

Thanks :)
